### PR TITLE
Separate atmosphere preprocessor

### PR DESCRIPTION
**Pull request by Schneegans**
This extracts the preprocessing code of the Bruneton atmosphere model into a separate binary. This allows us to precompute the scattering data offline and thus improve the startup time of CosmoScout VR significantly.

The precomputed data is stored in Tiff files and is available in [this repository](https://github.com/cosmoscout/atmosphere-data/tree/main) which is used as a submodule in CosmoScout VR.

### CODE REVIEW

**Comment by JonasGilg** [comment]
I get the following error, when I try to compile with Visual Studio 2019:



```bash
[...]/plugins/csp-atmospheres/src/models/bruneton/Model.cpp(65): error C2079: "file" verwendet undefiniertes class "std::basic_ifstream<char,std::char_traits<char>>"
[...]/plugins/csp-atmospheres/src/models/bruneton/Model.cpp(65): error C2440: "Initialisierung": "std::basic_string<char,std::char_traits<char>,std::allocator<char>>" kann nicht in "int" konvertiert werden
```

**Comment by Schneegans** [comment]
> I get the following error, when I try to compile with Visual Studio 2019:
> 
> ```shell
> [...]/plugins/csp-atmospheres/src/models/bruneton/Model.cpp(65): error C2079: "file" verwendet undefiniertes class "std::basic_ifstream<char,std::char_traits<char>>"
> [...]/plugins/csp-atmospheres/src/models/bruneton/Model.cpp(65): error C2440: "Initialisierung": "std::basic_string<char,std::char_traits<char>,std::allocator<char>>" kann nicht in "int" konvertiert werden
> ```

I fixed some other msvc-related issues. However, this error I do not get (also Visual Studio 2019) and the CI builds do not complain either. Do you have an idea what could be the reason?

**Comment by JonasGilg** [comment]
Okay, I got it to compile. If I start it now I get the following error, when the atmospheres plugin loads:

```
[I] csp-atmospheres ... Loading plugin...
../share/resources/atmosphere-data/earth/phase.tif: Deflate compression support is not configured.
../share/resources/atmosphere-data/earth/phase.tif: Deflate compression support is not configured.
../share/resources/atmosphere-data/earth/phase.tif: Deflate compression support is not configured.
../share/resources/atmosphere-data/earth/transmittance.tif: Deflate compression support is not configured.
../share/resources/atmosphere-data/earth/transmittance.tif: Deflate compression support is not configured.
../share/resources/atmosphere-data/earth/transmittance.tif: Deflate compression support is not configured.
```

**Comment by Schneegans** [comment]
Yeah, I simply use textures without compression now. Seems to be the easier approach, I guess.

**Comment by Schneegans** [comment]
You'll need a `git submodule update` to fetch the new textures.

**Review by JonasGilg — CHANGES_REQUESTED** [review]

File: plugins/csp-atmospheres/bruneton-preprocessor/Preprocessor.cpp [hunk 1]
```diff
@@ -675,35 +648,43 @@ std::string printAbsorbingComponent(Params::AbsorbingComponent const& component,
...
+    , mScatteringTextureDepth(mParams.mScatteringTextureRSize.get()) {
+
+  std::cout << "Preprocessing atmosphere..." << std::endl;
+
+  // Compute angular radius of the sun.
+  float sunRadius             = 696340000.F; // meters
+  mMetadata.mSunAngularRadius = std::asin(sunRadius / mParams.mSunDistance);
+
   // Compute the values for the SUN_RADIANCE_TO_LUMINANCE constant.
+  float sunAngularRadiusAtEarth = 0.0046547F; // radians
```
[the remark this answers is not in the document]
Comment (Schneegans) [reply plugins/csp-atmospheres/bruneton-preprocessor/Preprocessor.cpp:665]: This is used to compute the irradiance at the bodies position. `SOLAR_IRRADIANCE` is also hard-coded for Earth and together with the `sunAngularRadiusAtEarth`, it is corrected for any other position.

**Pull request merged**